src/mdns/mdns_mdnssd.cpp

```diff
@@ -340,6 +334,79 @@ void PublisherMDnsSd::Process(const MainloopContext &aMainloop)
     }
 }
 
+PublisherMDnsSd::DnssdServiceRegistration::~DnssdServiceRegistration(void)
+{
+    if (mServiceRef != nullptr)
+    {
+        DNSServiceRefDeallocate(mServiceRef);
+    }
+}
+
+PublisherMDnsSd::DnssdHostRegistration::~DnssdHostRegistration(void)
+{
+    VerifyOrExit(mServiceRef != nullptr && mRecordRef != nullptr);
+
+    if (IsCompleted())
+    {
+        // The Bonjour mDNSResponder somehow doesn't send goodbye message for the AAAA record when it is
+        // removed by `DNSServiceRemoveRecord`. Per RFC 6762, a goodbye message of a record sets its TTL
+        // to zero but the receiver should record the TTL of 1 and flushes the cache 1 second later. Here
+        // we remove the AAAA record after updating its TTL to 1 second. This has the same effect as
+        // sending a goodbye message.
+        // TODO: resolve the goodbye issue with Bonjour mDNSResponder.
+        int dnsError = DNSServiceUpdateRecord(mServiceRef, mRecordRef, kDNSServiceFlagsUnique, mAddress.size(),
+                                              mAddress.data(), /* ttl */ 1);
+        otbrLogWarning("Failed to send goodbye message for host %s: %s", MakeFullHostName(mName).c_str(),
+                       DNSErrorToString(dnsError));
+    }
+
+    DNSServiceRemoveRecord(mServiceRef, mRecordRef, /* flags */ 0);
+    // TODO: ?
+    // DNSRecordRefDeallocate(mRecordRef);
+
+exit:
+    return;
+}
+
+Publisher::ServiceRegistrationPtr PublisherMDnsSd::FindServiceRegistration(const DNSServiceRef &aServiceRef)
```

Comment:
Just a sudden thought: we may be able to put this method in the base class. For each derived `ServiceRegistration`, we should provide an interface for comparison. 

It may or may not be worth doing. No need to change it in this PR.

Reply:
We don't want to expose `DNSServiceRef` to the base class `ServiceRegistration`.